### PR TITLE
WIP: feat(nx): add generic task execution

### PR DESCRIPTION
**Pull request by FrozenPandaz**
# WIP

This PR is still in progress but below is the information for the intentions of the PR.

## Current Behavior (This is the behavior we have today, before the PR is merged)

Task execution is currently handled by `npm-run-all` and unconfigurable to the user. There are limitations to what can be done through `npm-run-all` under the hood.

## Expected Behavior (This is the new behavior we can expect after the PR is merged)

This PR enables configuration of custom `TasksRunners`. The _default_ TasksRunner right now still uses `npm-run-all`.

### TasksRunners

A TasksRunner is a function with the following signature:

```ts
function customTasksRunner(tasks: Task[], options: any): Observable<BuildEvent> {
  // Run Tasks
}
```

#### 🚀Remote Task Execution

The next step of this effort is that Nrwl is working on a task runner which will execute tasks remotely which will allow users to distribute their builds across multiple machines!

#### Ecosystem

This opens up an ecosystem of custom task runners. Users can create their own to suit their needs in general. We will also move the default task runner more extensible to allow an ecosystem to develop more easily.

### CODE REVIEW

File: packages/workspace/src/command-line/affected-apps.ts [hunk 1]
```diff
@@ -23,60 +23,83 @@ export type AffectedFetcher = (
   projects: ProjectNode[],
   dependencies: Deps,
   touchedProjects: string[]
-) => string[];
+) => AffectedProject[];
 
-export function affectedAppNames(
+export function affectedApps(
```
Comment (vsavkin) [inline packages/workspace/src/command-line/affected-apps.ts]: I'd like to change the return value type to somethign like this:

```
{
  projects: ProjectNode[], // all projects
  dependencies: Deps, // all deps
  affectedProjects: string[]
}
```

File: packages/workspace/src/task-runner/task-runner.ts [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+import { Target } from '@angular-devkit/architect';
+
+export interface Task {
+  id: string;
+  target: Target;
+  overrides: Object;
+  dependencies: string[];
+}
+
+export enum BuildEventType {
```
Comment (vsavkin) [inline packages/workspace/src/task-runner/task-runner.ts]: Rename into `EventType`. We do more than just build

File: packages/workspace/src/task-runner/task-runner.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+  TaskComplete = '[Task] Complete'
+}
+
+export interface BuildEvent {
```
Comment (vsavkin) [inline packages/workspace/src/task-runner/task-runner.ts]: Event or TaskEvent

**Comment by vsavkin** [comment]
A good way to express a graph is as follows:

```
type Node {id: string}

type Graph = {
  nodes: Node[],
  roots: string[],
  deps: {[id: string]: string[]}
}
```

You do something like this:

```
type Node {id: string, deps: Nodes[]}
type Graph = Node[]
```

or something like this:

```
type Node {id: string, deps: string[]}
type Graph = Node[]
```

In other words, a good way to express a graph is to separate nodes and edges. You may or may not include the information about the roots (you can always derive it). If you do this, you will never have the same information duplicated.

Your approch works OK when you have a tree or a forest.

File: packages/workspace/src/task-runner/default-task-runner.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        subscriber.complete();
+        // fix for https://github.com/nrwl/nx/issues/1666
+        if (process.stdin['unref']) (process.stdin as any).unref();
+      });
+  });
+};
+
+function getLegacyTaskOverrides(options: any) {
+  const legacyTaskOverrides = { ...options };
+  delete legacyTaskOverrides.maxParallel;
```
Comment (vsavkin) [inline packages/workspace/src/task-runner/default-task-runner.ts]: the other options (e.g., --head) are filtered out elsewhere. right?

File: packages/workspace/src/task-runner/default-task-runner.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    task.overrides = {
+      ...task.overrides,
+      ...additionalTaskOverrides
+    };
+  });
+  return new Observable(subscriber => {
+    const commands = getCommands(tasks);
+    runAll(commands, {
+      parallel: options.parallel || false,
+      maxParallel: options.maxParallel || 1,
```
Comment (vsavkin) [inline packages/workspace/src/task-runner/default-task-runner.ts]: it should be ||3

File: packages/workspace/src/command-line/shared.ts [hunk 6]
```diff
@@ -399,44 +412,60 @@ export const getAffected = (affectedNamesFetcher: AffectedFetcher) => (
...
+  return getAllProjects(
+    proj =>
+      proj.architect[target] &&
+      proj.architect[target].configurations &&
+      proj.architect[target].configurations[configuration]
+  );
 }
 
-export function getAllProjectsWithTarget(target: string) {
+export function getAllProjects(
```
Comment (vsavkin) [inline packages/workspace/src/command-line/shared.ts]: see the comment about the graph. I think we should change the return value

File: packages/workspace/src/command-line/shared.ts [hunk 7]
```diff
@@ -399,44 +412,60 @@ export const getAffected = (affectedNamesFetcher: AffectedFetcher) => (
...
-  return getProjectNames(p => p.architect[target]);
+export function getAllProjectsWithTargetAndConfiguration(
+  target: string,
+  configuration?: string
+): AffectedProject[] {
+  return getAllProjects(
+    proj =>
+      proj.architect[target] &&
+      proj.architect[target].configurations &&
+      proj.architect[target].configurations[configuration]
```
Comment (vsavkin) [inline packages/workspace/src/command-line/shared.ts]: this seems weird. it looks like configuration is optional but in this if statement it's used as if it wasn't

File: packages/workspace/src/command-line/shared.ts [hunk 8]
```diff
@@ -388,7 +401,7 @@ export function readNxJson(): NxJson {
 
 export const getAffected = (affectedNamesFetcher: AffectedFetcher) => (
   touchedFiles: string[]
-): string[] => {
+): AffectedProject[] => {
```
Comment (vsavkin) [inline packages/workspace/src/command-line/shared.ts]: see the commend about the graph.

I'd like all of these return the graph with "highlighted" nodes and the roots.

File: packages/workspace/src/command-line/nx-commands.ts [hunk 9]
```diff
@@ -244,6 +244,15 @@ function withAffectedOptions(yargs: yargs.Argv): yargs.Argv {
       coerce: parseCSV,
       default: []
     })
+    .options('runner', {
+      describe: 'This is the name of the tasks runner configured in nx.json',
+      type: 'string'
+    })
+    .options('configuration', {
```
Comment (vsavkin) [inline packages/workspace/src/command-line/nx-commands.ts]: just so I understand. We are able to pass the configuration right now, we just don't use it for filtering.

**Comment by vsavkin** [comment]
I think we should also add:
- unit tests for the default task runner (with lots of flags)
- e2e tests for the task runner (the custom one)

**Review by vsavkin — CHANGES_REQUESTED** [review]

File: packages/workspace/src/task-runner/task-runner.ts [hunk 10]
```diff
@@ -0,0 +1,28 @@
...
+  type: BuildEventType;
+}
+
+export interface TaskCompleteEvent extends BuildEvent {
+  type: BuildEventType.TaskComplete;
+  success: boolean;
+}
+
+export type TasksRunner = (
+  tasks: Task[],
```
Comment (vsavkin) [inline packages/workspace/src/task-runner/task-runner.ts]: Given my comments about the graph, we might change it to an object that looks like this:

```
{
  tasks: {id: string; target: Target, overrides}
  roots: string[]
  deps: {[id: string]: string[]}
}
```

File: packages/workspace/src/task-runner/default-task-runner.ts (hunk 4 above)
Comment (FrozenPandaz) [reply packages/workspace/src/task-runner/default-task-runner.ts]: Yup, they are filtered out within `affected.ts#processArgs`

File: packages/workspace/src/command-line/nx-commands.ts (hunk 9 above)
Comment (FrozenPandaz) [reply packages/workspace/src/command-line/nx-commands.ts]: Yup

**Comment by vsavkin** [comment]
You use both "task runner" and "tasks runner". Could you make sure that all file names and all variables use "tasks"?

**Pull request merged**